Add support for TRUNCATE USING TIMEOUT

bhalevy opened the pull request:
Extend the cql3 truncate statement to accept attributes,
similar to modification statements.

To achieve that we define cql3::statements::raw::truncate_statement
derived from raw::cf_statement, and implement its pure virtual
prepare() method to make a prepared truncate_statement.

The latter is no longer derived from raw::cf_statement,
and just stores a schema_ptr to get to the keyspace and column_family.

`test_truncate_using_timeout` cql-pytest was added to test
the new USING TIMEOUT feature.
    
Fixes #11408

Also, update docs/cql/ddl.rst truncate-statement section respectively.

bhalevy commented:
Cc @tzach @annastuchlik

nyh commented on cql3/statements/raw/truncate_statement.hh:
```diff
@@ -0,0 +1,42 @@
+    std::unique_ptr<attributes::raw> _attrs;
+public:
+    /**
+     * Creates a new UpdateStatement from a column family name, columns map, consistency
```
Copy-pasted comment is incorrect.
To be honest, I don't understand what is this new "raw" truncate statement and how it differs from the regular (cooked? :-)) truncate statement.

bhalevy replied:
The difference is in containing `attributes::raw` vs. prepared `attributes`
(and being derived from `raw::cf_statement` vs. keeping a `schema_ptr` member)

Will fix the comment nevertheless. thanks for spotting.

bhalevy replied:
fixed in v2

nyh commented on cql3/Cql.g:
```diff
@@ -44,7 +44,7 @@ options {
 #include "cql3/statements/drop_service_level_statement.hh"
 #include "cql3/statements/detach_service_level_statement.hh"
-#include "cql3/statements/truncate_statement.hh"
+#include "cql3/statements/raw/truncate_statement.hh"
```
I found that "the latter" is supposed to refer to truncate_statement, the last word on the previous sentence. This wasn't obvious at all in first reading.

bhalevy replied:
right, I'll rephrase to be clearer, hopefully :)

bhalevy replied:
done in v2

nyh reviewed (changes requested):
Looks good.
Please also update docs/cql/cql-extensions.md, which has a USING TIMEOUT section but is really confusing on which statement it applies to - it forgets to mention it applies to inserts (although there is an example of that in the end) and then says it "also" applies to SELECT. And now you're adding support for TRUNCATE. I think we should explicitly list the specific CQL requests to which we added the USING TIMEOUT feature.

bhalevy commented on cql3/Cql.g:
```diff
@@ -1058,10 +1058,18 @@ dropIndexStatement returns [std::unique_ptr<drop_index_statement> expr]
 /**
-  * TRUNCATE <CF>;
+  * TRUNCATE <CF>
+  * [USING TIMEOUT]
```
hmm, '<duration>' is more accurate

bhalevy commented on cql3/Cql.g:
```diff
@@ -1058,10 +1058,18 @@ dropIndexStatement returns [std::unique_ptr<drop_index_statement> expr]
+        auto attrs = std::make_unique<cql3::attributes::raw>();
+    }
+    : K_TRUNCATE (K_COLUMNFAMILY)? cf=columnFamilyName
+      ( usingClause[attrs] )?
```
There's no validation in this series so use of other `USING` attributes will succeed and they'll just be ignored.
As a follow up, I plan that after https://github.com/scylladb/scylladb/issues/11230 is done, the user could also use `TRUNCATE USING TIMESTAMP <timestamp>`, so it makes sense to set the stage for it and
maybe just add validation when preparing the truncate_statement internally from the raw statement.

bhalevy replied:
done in v2

avikivity replied:
TRUNCATE USING TIMESTAMP requires truncate tombstones, since there could be data after the timestamp.

I don't see how @nyh's comment was addressed.

bhalevy commented on cql3/statements/truncate_statement.cc:
```diff
@@ -8,32 +8,62 @@
+uint32_t truncate_statement::get_bound_terms() const
 {
-    return std::make_unique<prepared_statement>(::make_shared<truncate_statement>(*this));
+    return 0;
```
I've no idea, it's a pure virtual method of `class cql_statement` that we have to override.
I just kept the existing implementation (the diff patch obfuscates that)

bhalevy commented on test/cql-pytest/test_using_timeout.py:
```diff
@@ -152,3 +153,15 @@ def invalid(stmt):
+    key = unique_key_int()
+    cql.execute(f"INSERT INTO {table} (p,c,v) VALUES ({key},1,1) USING TIMEOUT 60m AND TTL 1000000 AND TIMESTAMP 321")
+    res = list(cql.execute(f"SELECT ttl(v), writetime(v) FROM {table} WHERE p = {key} and c = 1"))
+    assert len(res) == 1 and res[0].ttl_v > 0 and res[0].writetime_v == 321
```
yeah, copy-pasta from the test above.
I was lazy about that

bhalevy replied:
fixed in v2

bhalevy commented on test/cql-pytest/test_using_timeout.py:
```diff
@@ -152,3 +153,15 @@ def invalid(stmt):
+    res = list(cql.execute(f"SELECT ttl(v), writetime(v) FROM {table} WHERE p = {key} and c = 1"))
+    assert len(res) == 0
+    with pytest.raises(NoHostAvailable):
+        cql.execute(f"TRUNCATE TABLE {table} USING TIMEOUT 0s")
```
That's a good point.  In a sense, this is similar to writing with CL=ALL, so `NoHostAvailable` does makes sense to me.

However, in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/33/testReport/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split026___test_truncate_base_during_mv_insert_test_with_auto_snapshot/
I see that the dtest got this error:
```
cassandra.OperationTimedOut: errors={<Host: 127.0.52.3:9042 dc1>: <Error from server: code=1003 [Error during truncate] message="Error during truncate: seastar::rpc::timeout_error (rpc call timed out)">, <Host: 127.0.52.2:9042 dc1>: <Error from server: code=1003 [Error during truncate] message="Error during truncate: seastar::rpc::timeout_error (rpc call timed out)">}, last_host=127.0.52.1:9042
```

so I'll see if NoHostAvailable stemmed from setting the timeout to 0 in particular.
(since I didn't change anything on the timeout handling in the truncate path from the storage_proxy down)

nyh replied:
Does a write with CL=ALL which times out not return a WriteTimeout? If not, what is WriteTimeout (I just saw it's available in the driver's list of errors)? :-(

bhalevy replied:
I think the answer is: "it's complicated".
For example, see https://github.com/scylladb/scylla-dtest/blob/3bc3ea944c1c19ee12d707408ce757977948413a/tools/assertions.py#L74-L85

bhalevy replied:
The above example is just to show the proliferation of errors in cassandra.
(not directly related to the timeout case).

In any case, I think that the error type is orthogonal to this change and I suggest that we deal with it later, under a different GH issue.

nyh replied:
Ok.

bhalevy commented:
In v2 ([753f010](https://github.com/scylladb/scylladb/commit/753f010ef370eb2c9933adf39f5891a4031a9154)):
- clarified commit message
- updated docs/cql/cql-extensions.md
  - preliminary patch updates the documentation prior to adding support for TRUNCATE and is backportable
- update comment for truncateStatement in Cql.g
- added validation that the USING attribute do not specify TTL or TIMESTAMP
  - augmented cql-pytest to test that they generate invalid cql query exception
- fixed copy-paste typo in cql3::statements::raw::truncate_statement declaration.
- simplified `test_truncate_using_timeout` INSERT and SELECT queries

nyh commented on cql3/statements/truncate_statement.cc:
```diff
@@ -8,32 +8,71 @@
+    }
+    if (_attrs->timestamp) {
+        throw exceptions::invalid_request_exception("TRUNCATE does not support USING TIMESTAMP");
+    }
```
If we later add more of these USING options, this won't check them. Would have been nice if we could verify that **only** the timeout option is set. I'm not sure how. One way this could maybe be done is to create an attr object with *only* the timeout and then check if the two objects are equal (requires an operator== that works properly).

But maybe this code is good enough, and we can improve it later if we add more options.

avikivity replied:
I think it's better to restrict it in the grammar.

bhalevy replied:
ok. Will change for `selectStatement` too then.

bhalevy replied:
Ref: https://github.com/scylladb/scylladb/blob/e5e75ba43c79a575c44f9b0824374b2be0798b72/cql3/statements/select_statement.cc#L1573-L1580

bhalevy replied:
addressed in v3

nyh commented on test/cql-pytest/test_using_timeout.py:
```diff
@@ -152,3 +153,19 @@ def invalid(stmt):
+    table = table1
+    key = unique_key_int()
+    cql.execute(f"INSERT INTO {table} (p,c,v) VALUES ({key},1,1)")
+    res = list(cql.execute(f"SELECT ttl(v), writetime(v) FROM {table} WHERE p = {key} and c = 1"))
```
selecting ttl(v), writetime(v) in this test instead of just v is weird ;-) But it's not wrong and I guess you copied it from a different test.

bhalevy replied:
cleaned up in v3

nyh commented:
Hi @benipeled it seems that because this pull request includes, in addition to code, a change to a document, the CI thinks it only needs to do some documentation checks on it, and doesn't build or test it. Can you please check and fix this logic?
Also, how can I force the CI to check this PR even before the bot is fixed?

avikivity commented:
> 
> Also, update docs/cql/ddl.rst truncate-statement section respectively.


Yes!

nyh commented:
> Hi @benipeled it seems that because this pull request includes, in addition to code, a change to a document, the CI thinks it only needs to do some documentation checks on it, and doesn't build or test it. Can you please check and fix this logic? Also, how can I force the CI to check this PR even before the bot is fixed?

Now the CI build job started. I don't know if you did something to start it, or it just took a long time to start and I was just impatient. Sorry about the spam.

avikivity commented on docs/cql/ddl.rst:
```diff
@@ -961,12 +961,20 @@ A table can be truncated using the ``TRUNCATE`` statement:
 .. code-block::
    
    truncate_statement: TRUNCATE [ TABLE ] `table_name`
+                     : [ USING TIMEOUT `timeout` ]
```
Should specify the version that this was added in.

nyh replied:
The version of what - Scylla? Because we don't have official CQL versioning (see https://github.com/scylladb/scylladb/pull/11320).
I thought we were aiming for versioned documents, so if you look at the documentation of Scylla 5.0 you'll see only the features that Scylla 5.0 supports.

avikivity replied:
It's still good to know when something was added, so you don't get surprised if you run it on an older version and it stops working.

nyh replied:
Maybe, but then again - almost every feature we added at some point. Do we keep a version comment on every feature forever? Do we only only do this for features introduced in the last year (or whatever), and go back once in a while to remove archaic version information, or what?

avikivity replied:
Why not? e.g. cppreference keeps a version tag for everything. If it's unobtrusive, it isn't a problem.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1999/

bhalevy commented:
In v3 ([ba82932](https://github.com/scylladb/scylladb/commit/ba82932e4ea16986a31a3fa368f798c1ffdeeb6b)):
- text in cql-extensions USING TIMEOUT section further improved
- added "cql3: attributes: introduce mask and get_set()"
  - used in delete, select, and truncate statements to blanket-assert that only the particularly supported attributes are set.
- cql3: deleteStatement: restrict to USING TIMEOUT|TIMESTAMP in grammar
- cql3: selectStatement: restrict to USING TIMEOUT in grammar
- cql3, storage_proxy: add support for TRUNCATE USING TIMEOUT
  - use usingTimeoutClause in Cql.g, like selectStatement
  - simplified SELECT query in `test_truncate_using_timeout`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2013/

nyh commented on cql3/attributes.cc:
```diff
@@ -149,4 +149,26 @@ lw_shared_ptr<column_specification> attributes::raw::timeout_receiver(const sstr
     return make_lw_shared<column_specification>(ks_name, cf_name, ::make_shared<column_identifier>("[timeout]", true), duration_type);
 }
 
+static unsigned make_bit_mask(bool timestamp, bool time_to_live, bool timeout) noexcept {
```
And I can't be sure without trying this myself, but looks too me like a very verbose patch just to do run-time validation. I wonder if it's possible to do what I proposed in the beginning - use a defaulted operator== (does this exit?) to check if the attributes are equal to ones with just the allowed attributes.

bhalevy replied:
A customized operator== will need to compare the presence of the different attributes rather than their values.
I think it's an overkill.
The runtime check prevents against programming errors, not user errors, with the syntax-level checking.
But since we're only dealing with those 3 kinds of attributes it might be an overkill indeed.

bhalevy replied:
bit_masks patch dropped in v4

nyh commented on cql3/Cql.g:
```diff
@@ -518,6 +518,15 @@ usingClauseObjective[std::unique_ptr<cql3::attributes::raw>& attrs]
+
+usingClauseNoTTLObjective[std::unique_ptr<cql3::attributes::raw>& attrs]
+    : K_TIMESTAMP ts=intValue { attrs->timestamp = ts; }
+    | K_TIMEOUT to=term { attrs->timeout = to; }
```
I wonder (but maybe it's not important to think about this right now) whether it's better to call this "usingClauseNoTTL" or "UsingClauseTimestampOrTimeout". If we later add a new "USING XYZ", are we more likely to want to allow it in this one or not?

bhalevy replied:
I'm not sure at this point.
Like I said above, the only attributes applicable to writes are TIMESTAMP and TTL.
TTL itself is specific to INSERT/UPDATE that may set the TTL on the upserted entity
and `usingClauseNoTTL` makes sense for delete-like queries such as DELETE and TRUNCATE,
for which both the TIMESTAMP and TIMEOUT are applicable.
Maybe calling it as you suggested is the right thing to do at the moment.

avikivity replied:
Maybe we should have a Santa Clause

nyh commented on cql3/Cql.g:
```diff
@@ -555,6 +564,7 @@ updateConditions returns [conditions_type conditions]
  * USING TIMESTAMP <long>
  * WHERE KEY = keyname
    [IF (EXISTS | name = value, ...)];
+ * [USING (TIMESTAMP <timestamp> | TIMEOUT <duration>) [AND ...]]
```
Above we have `USING TIMESTAMP <long>`, here `TIMESTAMP <timestamp>` - is the difference intentional? If not, we should probably use the same one.

bhalevy replied:
`long` is more correct as the timestamp value is indeed a pure number and not a cql timestamp (i.e. millis since epoch), thanks.

bhalevy replied:
fixed in v4

nyh commented on cql3/statements/delete_statement.cc:
```diff
@@ -99,9 +99,6 @@ delete_statement::delete_statement(cf_name name,
     , _deletions(std::move(deletions))
     , _where_clause(std::move(where_clause))
 {
-    if (_attrs->time_to_live) {
```
If you removed this, you probably don't need the mask machinary you added in the previous patch.

bhalevy replied:
changed to assert in v4

bhalevy commented on docs/cql/cql-extensions.md:
```diff
@@ -48,10 +48,16 @@ way as other options by using `WITH` clause:
+INSERT, UPDATE, DELETE, PRUNE MATERIALIZED VIEW, BATCH.
+
+In order for this parameter to be effective for read operations as well,
+the USING TIMEOUT clause may be specified for SELECT statements.
```
TIMEOUT is applicable to all kind of queries.
TIMESTAMP and TTL are inherently applicable only to modification statements.
(We could in theory support timestamp-consistent SELECT USING TIMESTAMP that will ignore updates after the given timestamp but it's not planned in the foreseeable future)

bhalevy replied:
simplified in v4 (but still separated based on the query kind)

bhalevy commented:
In v4 ([b178813](https://github.com/scylladb/scylladb/commit/b178813cba527f15757802e74e5b8a8519481f33)):
- dropped "docs: cql-extensions: elaborate on USING TIMEOUT section."
- dropped "cql3: attributes: introduce mask and get_set()"
  - changed attr validation to assert when syntax is enforced in Cql.g
- docs: cql-extensions: add TRUNCATE to USING TIMEOUT section.
  - simplified based on @nyh review comments
- `USING TIMESTAMP <timestamp>` comment in Cql.g changed to `USING TIMESTAMP <long>`
  - and cleaned up other related changes / additions in Cql.g comments.

bhalevy commented on docs/cql/cql-extensions.md:
```diff
@@ -48,10 +48,16 @@ way as other options by using `WITH` clause:
-to attach USING clause to SELECT statements.
+The parameter acts as part of the USING clause, and thus can usually be combined with other
+parameters - like timestamp and time-to-live, with some restrictions for particular queries
+like DELETE, that does not support USING TTL, or SELECT, that is read-only so it
```
done in v4

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2397/

annastuchlik commented:
@tzach Should this PR be backported to branch 5.1?

annastuchlik commented:
@bhalevy Thanks, I'm trying to make sure if anything is missing from the docs for 5.1. Everything is clear to me now, as this PR includes both the change and the docs.

tzach commented:
> as this PR includes both the change and the docs

@bhalevy set a good example!
